ocaml/xapi/message_forwarding.ml

```diff
@@ -1687,9 +1687,16 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
             (* Intra-pool: reserve resources on the destination host, then
              * forward the call to the source. *)
             let snapshot = Db.VM.get_record ~__context ~self:vm in
-            (fun ~local_fn ~__context ~vm op ->
-               allocate_vm_to_host ~__context ~vm ~host ~snapshot ~host_op:`vm_migrate ();
-               forward_vm_op ~local_fn ~__context ~vm op)
+            let clear_migrate_op () =
+              Helpers.with_global_lock
+                (fun () ->
+                  finally_clear_host_operation ~__context ~host ~host_op:`vm_migrate ();
+                  clear_scheduled_to_be_resident_on ~__context ~vm) in
```

Comment:
`finally_clear_host_operation` fixes the issue, but `clear_scheduled_to_be_resident_on` also looks important in the case that there is a vgpu

Reply:
Is there any reason to call the db-changing functions here isntead of inside `allocate_vm_to_host`? It seems strange that it's done here and not where in the function where the DB is changed.

Reply:
you don't want to do it in there, because it does not give the right behaviour - you want to wait till the op has finished, i.e. after `foward_vm_op`